### PR TITLE
Sdpa fusion gpt oss

### PR DESCRIPTION
**Pull request by CuriousPanCake**
### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

**Comment by yeonbok** [comment]
Could you add a decompose pass too ?

**Review by yeonbok — APPROVED** [review]

File: src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp [hunk 1]
```diff
@@ -89,6 +94,8 @@ bool SDPAFusion::run_on_model(const std::shared_ptr<ov::Model>& model) {
     ov::pass::SymbolicOptimizations symbolic_optimizations(false, get_pass_config());
     auto symbolic_ctx_manager = symbolic_optimizations.get_manager();
     symbolic_ctx_manager->register_pass<ov::pass::SDPAFusionMatcher>();
+    symbolic_ctx_manager->register_pass<ov::pass::SDPAFusionMatcherSinks>();
+    symbolic_ctx_manager->register_pass<ov::pass::Serialize>(std::string("after_SDPAFusionMatcherSinks.xml"), std::string("after_SDPAFusionMatcherSinks.bin"));
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp]:
```suggestion
```

File: src/common/transformations/include/transformations/common_optimizations/sdpa_fusion.hpp [hunk 2]
```diff
@@ -62,6 +62,12 @@ class TRANSFORMATIONS_API SDPAReshapeFusion : public ov::pass::MatcherPass {
     SDPAReshapeFusion();
 };
 
+class TRANSFORMATIONS_API SDPAFusionMatcherSinks : public ov::pass::MatcherPass {
```
Comment (rkazants) [inline src/common/transformations/include/transformations/common_optimizations/sdpa_fusion.hpp:118]: can you please give a picture of the pattern to be fused

File: src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp [hunk 3]
```diff
@@ -419,5 +425,194 @@ SDPAFusionMatcher::SDPAFusionMatcher() {
...
+    MATCHER_SCOPE(SDPAFusionMatcherSinks);
+
+    using namespace ov::op;
+    using namespace ov::pass::pattern;
+
+    auto v = any_input(rank_equals(4));
+    auto k = any_input(rank_equals(4));
+    auto q = any_input(rank_equals(4));
+
+
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp]:
```suggestion
```

File: src/core/tests/type_prop/scaled_dot_product_attention.cpp [hunk 4]
```diff
@@ -248,6 +248,22 @@ TEST(type_prop, scaled_dot_product_attention_infer_5_dynamic_attn_partial) {
     EXPECT_EQ(op->get_output_partial_shape(0), (dynamic));
 }
 
+TEST(type_prop, scaled_dot_product_attention_infer_6_dynamic) {
```
Comment (rkazants) [inline src/core/tests/type_prop/scaled_dot_product_attention.cpp]: is there any transformation test?

Comment (CuriousPanCake) [reply src/core/tests/type_prop/scaled_dot_product_attention.cpp]: Added

File: src/common/transformations/include/transformations/common_optimizations/sdpa_fusion.hpp (hunk 2 above)
Comment (CuriousPanCake) [reply src/common/transformations/include/transformations/common_optimizations/sdpa_fusion.hpp:118]: There're a lot of SDPA variations to fuse, so listing all of them is not feasible. Added as screenshot to the PR

File: src/common/transformations/src/transformations/op_conversions/scaled_dot_product_attention_decomposition.cpp [hunk 5]
```diff
@@ -91,6 +93,7 @@ std::shared_ptr<ov::Node> ov::pass::ScaledDotProductAttentionDecomposition::deco
     auto query = node->input_value(0);
     auto key = node->input_value(1);
     auto value = node->input_value(2);
+    std::cout << "query: " << query << std::endl;
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/op_conversions/scaled_dot_product_attention_decomposition.cpp]:
```suggestion
```

File: src/common/transformations/tests/common_optimizations/sdpa_fusion_test.cpp [hunk 6]
```diff
@@ -59,6 +62,19 @@ class SDPA {
         m_scale = new_scale;
     }
 
+    void set_sinks(const PartialShape& sinks_shape, const std::vector<size_t>& sinks_broadcast_shape) {
+        // Adjust the code for sinks if you see other values for them.
+        // For now, there has been only one model with sinks: gpt-oss
+        with_sinks = true;
+        std::vector<int> broadcast_shape_value(sinks_broadcast_shape.begin(), sinks_broadcast_shape.end());
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/common_optimizations/sdpa_fusion_test.cpp]:
```suggestion
        std::vector<size_t> broadcast_shape_value(sinks_broadcast_shape.begin(), sinks_broadcast_shape.end());
```

File: src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp [hunk 7]
```diff
@@ -419,5 +425,202 @@ SDPAFusionMatcher::SDPAFusionMatcher() {
...
+                    auto axes_const = v0::Constant::create(ov::element::i64, ov::Shape{axes.size()}, axes);
+                    auto mask_squeeze = std::make_shared<v0::Squeeze>(mask_input, axes_const);
+                    mask_squeeze->set_friendly_name(mask->get_friendly_name());
+                    ov::copy_runtime_info(m.get_matched_nodes(), mask_squeeze);
+                    mask_input = mask_squeeze;
+                }
+            }
+        } else {
+            mask_input = v0::Constant::create(T, ov::Shape{}, {0});
+        }
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp:563]: Would be good to put the common code in some helper functions to reuse them between the fusions.
This block of code seems to be a duplication of the common SDPAFusionMatcher code from:
https://github.com/openvinotoolkit/openvino/blob/d1433f05e8d06259b7c26d2034d52f4ea392fcd4/src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp#L348-L378

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp [hunk 8]
```diff
@@ -329,8 +334,70 @@ const std::shared_ptr<ov::Node> scaled_dot_product_attention_decomposition(std::
...
+        manager.register_pass<ov::pass::ScaledDotProductAttentionDecomposition>();
+        manager.register_pass<ov::pass::VisualizeTree>("after.svg");
+    }
+
+    {
+        auto ref =
+            scaled_dot_product_attention_decomposition(query, key, value, attention_mask, scale, casual, false, sinks);
+        model_ref = std::make_shared<ov::Model>(OutputVector{ref},
+                                                ParameterVector{query, key, value, attention_mask, scale, sinks});
+        ov::pass::VisualizeTree("model_ref.svg").run_on_model(model_ref);
```
Comment (mitruska) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]: Is the VisualizeTree pass here a debug leftover?
It will generate image with transformations unit test run by default (should be rather conditional or removed).

**Review by mitruska — APPROVED** [review]
The patterns LGTM. Would be good to refactor for reusage of the common code part with existing SDPAFusion (not a blocker, possible as follow up improvement). 
Minor comment for graph visualization in the tests.

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp (hunk 8 above)
Comment (CuriousPanCake) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]:
```suggestion
```

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp [hunk 9]
```diff
@@ -329,8 +334,70 @@ const std::shared_ptr<ov::Node> scaled_dot_product_attention_decomposition(std::
...
+                                                                             value,
+                                                                             attention_mask,
+                                                                             scale,
+                                                                             sinks,
+                                                                             casual);
+        model = std::make_shared<ov::Model>(OutputVector{sdpa},
+                                            ParameterVector{query, key, value, attention_mask, scale, sinks});
+        manager.register_pass<ov::pass::VisualizeTree>("before.svg");
+        manager.register_pass<ov::pass::ScaledDotProductAttentionDecomposition>();
+        manager.register_pass<ov::pass::VisualizeTree>("after.svg");
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]:
```suggestion
```

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp [hunk 10]
```diff
@@ -329,8 +334,70 @@ const std::shared_ptr<ov::Node> scaled_dot_product_attention_decomposition(std::
...
+        auto sdpa = std::make_shared<ov::op::v13::ScaledDotProductAttention>(query,
+                                                                             key,
+                                                                             value,
+                                                                             attention_mask,
+                                                                             scale,
+                                                                             sinks,
+                                                                             casual);
+        model = std::make_shared<ov::Model>(OutputVector{sdpa},
+                                            ParameterVector{query, key, value, attention_mask, scale, sinks});
+        manager.register_pass<ov::pass::VisualizeTree>("before.svg");
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]:
```suggestion
```

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp (hunk 8 above)
Comment (CuriousPanCake) [reply src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]: Removed

File: src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp (hunk 7 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp:563]: Yes, for now we've reused the callback because of pattern complications. We'll integrate it in the general SDPAFusionMatcher in the next PR

File: src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp [hunk 11]
```diff
@@ -419,5 +425,202 @@ SDPAFusionMatcher::SDPAFusionMatcher() {
...
+    auto sinks_broadcast = wrap_type<v3::Broadcast>({sinks, any_input()});
+    auto sinks_concat = wrap_type<v0::Concat>({opt_mask_add, sinks_broadcast});
+    auto sinks_rm = wrap_type<v1::ReduceMax>({sinks_concat, any_input()});
+    auto sinks_sub = wrap_type<v1::Subtract>({sinks_concat, sinks_rm});
+
+    // Softmax axis can be:
+    // Pattern 1: axis = -1 (last axis)
+    // Pattern 2: axis = rank size - 1 (also means last axis for static rank inputs)
+    auto axis_predicate = ([](const ov::Output<ov::Node>& node) {
+        auto softmax = std::dynamic_pointer_cast<ov::op::v8::Softmax>(node.get_node_shared_ptr());
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp]: afaik, we need to use ov::as_type_ptr. This is important for OV.

File: src/common/transformations/include/transformations/common_optimizations/sdpa_fusion.hpp (hunk 2 above)
Comment (rkazants) [reply src/common/transformations/include/transformations/common_optimizations/sdpa_fusion.hpp:118]: Try to share some particular case for it and give a note what other variations can be. Will be much helpful.

File: src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp [hunk 12]
```diff
@@ -419,5 +425,202 @@ SDPAFusionMatcher::SDPAFusionMatcher() {
...
+        }
+
+        bool mask_present = pm.count(mask);
+        bool matmul_trasposes_k = pm.count(qk_transpose_b);
+
+        auto q_node = pm.at(q);
+        auto k_node = pm.at(k);
+        auto v_node = pm.at(v);
+
+        if (pm.at(mask).get_partial_shape().size() > 4) {
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp]: not clear. Is it a rank you want to compare? Is it static?

Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp]: We're comparing the rank here, changed.
Yeah, it's static, we check this above using the pattern predicate: `auto mask = any_input(has_static_rank());`

File: src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp (hunk 11 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp]: From what I heard, there were new findings that using `std::dynamic_pointer_cast` is also safe now, but I'll change to `ov::as_type_ptr`, no problem

File: src/common/transformations/include/transformations/common_optimizations/sdpa_fusion.hpp (hunk 2 above)
Comment (CuriousPanCake) [reply src/common/transformations/include/transformations/common_optimizations/sdpa_fusion.hpp:118]: Added ascii graph

File: src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp (hunk 11 above)
Comment (rkazants) [reply src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp]: Who said it?

Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp]: To be honest, don't remember right now :)
